daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/RollbackTest.scala

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package speedy
+
+import com.daml.lf.PureCompiledPackages
+import com.daml.lf.data.ImmArray
+import com.daml.lf.data.Ref.Party
+import com.daml.lf.language.Ast._
+import com.daml.lf.speedy.Compiler.FullStackTrace
+import com.daml.lf.speedy.SResult.SResultScenarioCommit
+import com.daml.lf.testing.parser.Implicits._
+import com.daml.lf.transaction.Node
+import com.daml.lf.transaction.SubmittedTransaction
+import com.daml.lf.validation.Validation
+import com.daml.lf.value.Value
+import com.daml.lf.value.Value._
+
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.prop.TableDrivenPropertyChecks
+import org.scalatest.wordspec.AnyWordSpec
+
+class ExceptionTest extends AnyWordSpec with Matchers with TableDrivenPropertyChecks {
+
+  private def typeAndCompile(pkg: Package): PureCompiledPackages = {
+    val rawPkgs = Map(defaultParserParameters.defaultPackageId -> pkg)
+    Validation.checkPackage(rawPkgs, defaultParserParameters.defaultPackageId, pkg)
+    data.assertRight(
+      PureCompiledPackages(rawPkgs, Compiler.Config.Default.copy(stacktracing = FullStackTrace))
+    )
+  }
+
+  private def runUpdateExpr_GetTx(pkgs1: PureCompiledPackages)(e: Expr): SubmittedTransaction = {
+    def transactionSeed: crypto.Hash = crypto.Hash.hashPrivateKey("RollbackTest.scala")
+    val m = Speedy.Machine.fromScenarioExpr(pkgs1, transactionSeed, e)
+    val res = m.run()
+    res match {
+      case SResultScenarioCommit(_, tx, _, _) =>
+        tx
+      case _ =>
+        sys.error(s"unexpected res: $res")
+    }
+  }
+
+  val pkgs: PureCompiledPackages = typeAndCompile(p"""
+      module M {
+
+        record @serializable MyException = { message: Text } ;
+
+        exception MyException = {
+          message \(e: M:MyException) -> M:MyException {message} e
+        };
+
+        record @serializable T1 = { party: Party, info: Int64 } ;
+        template (record : T1) = {
+          precondition True,
+          signatories Cons @Party [(M:T1 {party} record)] (Nil @Party),
+          observers Nil @Party,
+          agreement "Agreement",
+          choices {
+          }
+        };
+
+        val create0 : Party -> Scenario Unit = \(party: Party) ->
```

Comment:
is there actually a reason why any of these have to go via scenarios? Given that we’d like to get rid of them evenutally, I’d rather avoid building more tests on top of them. You should be able to just call update expressions directly and afterwards call `ptx.finish`

Reply:
In general,  I prefer our first level of test depend on the compiler. when test depend on the compiler: 
1 - It make debugging much more involve 
2 - You do not control as well the LF produces, it is maybe not an issue for you compiler guys that have a clear idea how this are compile, but it is not for me. 
3 - A test that tackles a particular feature will survive even if the compiler does not handle it anymore. 

To sum up, in my point of view we should have unit test that do not depends on the compiler **and** integration test.

For sure, we should not rewrite this test now, since it rely on invariant over the cid seeding that will very problably go away.

Reply:
I guess not. It was just convenient. I'll try and change to avoid Scenarios.

Reply:
I aggree with @moritzkiefer-da. 
However I will not rewrite them now, as the invariants (on cid seeding) they are based will probably go away soon. 
Let's postpone that until we produce actual rollback node.

Reply:
We don’t rely on contract ids in the test do we? We rely on the integer value in the contract which should be independent of any changes here.

Reply:
@cocreature. You are right.

Reply:
thanks for making the change!

Reply:
I didn't see this conversation, and already changed to avoid Scenarios.